tasks/write-to-a-windows-event-log/src/main.rs

```diff
@@ -0,0 +1,44 @@
+#[cfg(target_os = "windows")]
+mod bindings {
+    ::windows::include_bindings!();
+}
+
+const TOKEN_READ: u32 = 0x00020008;
+const EVENTLOG_WARNING_TYPE: u16 = 2;
+
+#[cfg(target_os = "windows")]
+fn main() {
+    use bindings::windows::win32::security::*;
+    use bindings::windows::win32::system_services::*;
+
+    unsafe {
+        let ph = GetCurrentProcess();
+        let mut th: HANDLE = HANDLE(0);
+        OpenProcessToken(ph, TOKEN_READ, &mut th);
+
+        let mut ti = std::ptr::null_mut();
+        let mut length: u32 = 0;
+        GetTokenInformation(th, TOKEN_INFORMATION_CLASS::TokenUser, ti, 0, &mut length);
```

Comment:
This API usage is incorrect. We need to first query how much buffer is needed, and then call the API again with an allocated buffer. Something like:

```
    // Determine the required buffer size.
    let mut length = 0u32;
    unsafe {
        GetTokenInformation(
            th,
            TOKEN_INFORMATION_CLASS::TokenUser,
            ptr::null_mut(),
            0,
            &mut length,
        )
    }
    .ok()
    .unwrap_err();

    // Retrieve the user token.
    let mut token_user_bytes = vec![0u8; usize::try_from(length).unwrap()];
    unsafe {
        GetTokenInformation(
            th,
            TOKEN_INFORMATION_CLASS::TokenUser,
            token_user_bytes.as_mut_ptr().cast(),
            length,
            &mut length,
        )
    }
    .ok()?;

    // Extract the user SID.
    let user_sid: PSID = unsafe { (*token_user_bytes.as_ptr().cast::<TOKEN_USER>()).User.Sid };

```

Reply:
It took me a while to figure out why the `unwrap_err() `is needed. But then I saw that no matter what the `GetTokenInformation()` will return `ERROR_INSUFFICIENT_BUFFER `when you call it for the first time to get the size of the buffer.